pkg/operator/staticpod/controller/staticpodstate/staticpodstate_controller.go

```diff
@@ -0,0 +1,166 @@
+	"github.com/openshift/library-go/pkg/operator/v1helpers"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/informers"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+)
+
+var (
+	staticPodStateControllerFailing      = "StaticPodsFailing"
+	staticPodStateControllerWorkQueueKey = "key"
+)
+
+// StaticPodStateController is a controller that watches the currentRevision and targetRevision fields for each node and spawn
+// installer pods to update the static pods on the master nodes.
+type StaticPodStateController struct {
+	targetNamespace, staticPodName string
+
+	operatorConfigClient v1helpers.StaticPodOperatorClient
+	kubeClient           kubernetes.Interface
+	eventRecorder        events.Recorder
+
+	// queue only ever has one item, but it has nice error handling backoff/retry semantics
+	queue workqueue.RateLimitingInterface
+}
+
+// NewStaticPodStateController creates a controller that watches static pods and will produce a failing status if the
+// static pods start crashing for some reason.
+func NewStaticPodStateController(
+	targetNamespace, staticPodName string,
+	kubeInformersForTargetNamespace informers.SharedInformerFactory,
+	operatorConfigClient v1helpers.StaticPodOperatorClient,
+	kubeClient kubernetes.Interface,
+	eventRecorder events.Recorder,
+) *StaticPodStateController {
+	c := &StaticPodStateController{
+		targetNamespace: targetNamespace,
+		staticPodName:   staticPodName,
+
+		operatorConfigClient: operatorConfigClient,
+		kubeClient:           kubeClient,
+		eventRecorder:        eventRecorder,
+
+		queue: workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "StaticPodStateController"),
+	}
+
+	operatorConfigClient.Informer().AddEventHandler(c.eventHandler())
+	kubeInformersForTargetNamespace.Core().V1().Pods().Informer().AddEventHandler(c.eventHandler())
+
+	return c
+}
+
+func (c *StaticPodStateController) sync() error {
+	operatorSpec, originalOperatorStatus, _, err := c.operatorConfigClient.GetStaticPodOperatorState()
+	if err != nil {
+		return err
+	}
+
+	switch operatorSpec.ManagementState {
+	case operatorv1.Unmanaged:
+		return nil
+	case operatorv1.Removed:
+		// TODO probably just fail.  Static pod managers can't be removed.
+		return nil
+	}
+
+	errs := []error{}
+	for _, node := range originalOperatorStatus.NodeStatuses {
+		pod, err := c.kubeClient.CoreV1().Pods(c.targetNamespace).Get(mirrorPodNameForNode(c.staticPodName, node.NodeName), metav1.GetOptions{})
+		if err != nil {
+			errs = append(errs, err)
+		}
+
+		for _, containerStatus := range pod.Status.ContainerStatuses {
+			if !containerStatus.Ready {
+				errs = append(errs, fmt.Errorf("nodes/%s pods/%s container=%q is not ready", node.NodeName, pod.Name, containerStatus.Name))
+			}
+			if containerStatus.State.Waiting != nil {
```

Comment:
not ready && waiting? do we want to report both?

Reply:
> not ready && waiting? do we want to report both? (at the same time)

I think so, they have different levels of detail and I'm not sure they're quite the same thing.